src/mount.ts

```diff
@@ -42,35 +47,127 @@ interface MountingOptions<Props> {
   shallow?: boolean
 }
 
-// TODO improve the typings of the overloads
-
-type ExtractComponent<T> = T extends { new (): infer PublicInstance }
-  ? PublicInstance
-  : any
+export type ComputedOptions = Record<
+  string,
+  ((ctx?: any) => any) | WritableComputedOptions<any>
+>
+export type ObjectEmitsOptions = Record<
+  string,
+  ((...args: any[]) => any) | null
+>
+export type EmitsOptions = ObjectEmitsOptions | string[]
 
 // Component declared with defineComponent
 export function mount<TestedComponent extends ComponentPublicInstance>(
   originalComponent: { new (): TestedComponent } & Component,
-  options?: MountingOptions<TestedComponent['$props']>
+  options?: MountingOptions<TestedComponent['$props'], TestedComponent['$data']>
 ): VueWrapper<TestedComponent>
-// Component declared with { props: { ... } }
-export function mount<TestedComponent extends ComponentOptionsWithObjectProps>(
-  originalComponent: TestedComponent,
-  options?: MountingOptions<ExtractPropTypes<TestedComponent['props'], false>>
-): VueWrapper<ExtractComponent<TestedComponent>>
-// Component declared with { props: [] }
-export function mount<TestedComponent extends ComponentOptionsWithArrayProps>(
-  originalComponent: TestedComponent,
-  options?: MountingOptions<Record<string, any>>
-): VueWrapper<ExtractComponent<TestedComponent>>
+
+// Functional Component declared with (props, ctx)=> render
+export function mount<Props, RawBindings = object>(
+  setup: (
+    props: Readonly<Props>,
+    ctx: SetupContext
+  ) => RawBindings | RenderFunction,
+
+  options: MountingOptions<Props, {}>
+): VueWrapper<
+  ComponentPublicInstance<
+    Props,
+    RawBindings,
+    {},
+    {},
+    {},
+    // public props
+    VNodeProps & Props
+  >
+>
+
 // Component declared with no props
 export function mount<
-  TestedComponent extends ComponentOptionsWithoutProps,
-  ComponentT extends ComponentOptionsWithoutProps & {}
+  Props = {},
+  RawBindings = {},
+  D = {},
+  C extends ComputedOptions = {},
+  M extends Record<string, Function> = {},
+  E extends EmitsOptions = Record<string, any>,
+  EE extends string = string
```

Comment:
Same here: if we really need all the generics, it would be great to have a test showcasing why.
Also, maybe pick more explicit names for the generic types and add comments: I struggle to follow what they refer to

Reply:
this is basically a `copy&paste` from [defineComponent.ts](https://github.com/vuejs/vue-next/blob/master/packages/runtime-core/src/apiDefineComponent.ts#L21-L126)

To make it simpler to maintain (bring the types from that file), I can add a comment explaining.